libbeat/statestore/backend/es/store.go

```diff
@@ -0,0 +1,340 @@
+)
+
+// The current typical usage of the state storage is such that the consumer
+// of the storage fetches all the keys and caches them at the start of the beat.
+// Then the key value gets updated (Set is called) possibly frequently, so we want these operations to happen fairly fast
+// and not block waiting on Elasticsearch refresh, thus the slight trade-off for performance instead of consistency.
+// The value is not normally retrieved after a modification, so the inconsistency (potential refresh delay) is acceptable for our use cases.
+//
+// If consistency becomes a strict requirement, the storage would need to implement possibly some caching mechanism
+// that would guarantee the consistency between Set/Remove/Get/Each operations at least for a given "in-memory" instance of the storage.
+
+type store struct {
+	ctx      context.Context
+	cn       context.CancelFunc
+	log      *logp.Logger
+	name     string
+	index    string
+	notifier *Notifier
+
+	chReady chan struct{}
+	once    sync.Once
+
+	mx     sync.Mutex
+	cli    *eslegclient.Connection
+	cliErr error
+}
+
+const docType = "_doc"
+
+func openStore(ctx context.Context, log *logp.Logger, name string, notifier *Notifier) (*store, error) {
+	ctx, cn := context.WithCancel(ctx)
+	s := &store{
+		ctx:      ctx,
+		cn:       cn,
+		log:      log.With("name", name).With("backend", "elasticsearch"),
+		name:     name,
+		index:    renderIndexName(name),
+		notifier: notifier,
+		chReady:  make(chan struct{}),
+	}
+
+	chCfg := make(chan *conf.C)
+
+	unsubFn := s.notifier.Subscribe(func(c *conf.C) {
+		select {
+		case chCfg <- c:
+		case <-ctx.Done():
+		}
+	})
+
+	go s.loop(ctx, cn, unsubFn, chCfg)
+
+	return s, nil
+}
+
+func renderIndexName(name string) string {
+	return "agentless-state-" + name
+}
+
+func (s *store) waitReady() error {
+	select {
+	case <-s.ctx.Done():
+		return s.ctx.Err()
+	case <-s.chReady:
+		return s.cliErr
+	}
+}
+
+func (s *store) SetID(id string) {
+	s.mx.Lock()
+	defer s.mx.Unlock()
+
+	if id == "" {
+		return
+	}
+	s.index = renderIndexName(id)
+}
+
+func (s *store) Close() error {
+	s.mx.Lock()
+	defer s.mx.Unlock()
+
+	if s.cn != nil {
+		s.cn()
+	}
+
+	if s.cli != nil {
+		err := s.cli.Close()
+		s.cli = nil
+		return err
+	}
+	return nil
+}
+
+func (s *store) Has(key string) (bool, error) {
+	if err := s.waitReady(); err != nil {
+		return false, err
+	}
+	s.mx.Lock()
+	defer s.mx.Unlock()
+
+	var v interface{}
+	err := s.get(key, v)
+	if err != nil {
+		if errors.Is(err, ErrKeyUnknown) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (s *store) Get(key string, to interface{}) error {
+	if err := s.waitReady(); err != nil {
+		return err
+	}
+	s.mx.Lock()
+	defer s.mx.Unlock()
+
+	return s.get(key, to)
+}
+
+func (s *store) get(key string, to interface{}) error {
+	status, data, err := s.cli.Request("GET", fmt.Sprintf("/%s/%s/%s", s.index, docType, url.QueryEscape(key)), "", nil, nil)
```

Comment:
We probably want to instrument this function with its own metrics, so we can observe the frequency of 429s, or just instrument it with tracing since that'll always be available where we plan to use it. There needs to be a way to tell if this is working properly.

This can be done as a follow up to this PR however.